pandas/core/arrays/interval.py

```diff
@@ -788,6 +789,25 @@ def size(self) -> int:
         # Avoid materializing self.values
         return self.left.size
 
+    def shift(self, periods: int = 1, fill_value: object = None) -> ABCExtensionArray:
+        if not len(self) or periods == 0:
+            return self.copy()
+
+        if isna(fill_value):
+            fill_value = self.dtype.na_value
+
+        # Note: The only change from ExtensionArray.shift is the lack of
+        # dtype=self.dtype here, since the shifted dtype may not match the
+        # original dtype.
+        empty = self._from_sequence([fill_value] * min(abs(periods), len(self)))
```

Comment:
It looks like we can get the job done by defining `empty` as:
```python
empty_len = min(abs(periods), len(self))
if isna(fill_value):
    fill_value = self.left._na_value
    empty = IntervalArray.from_breaks([fill_value] * (empty_len + 1))
else:
    empty = self._from_sequence([fill_value] * empty_len)
```
This also fixes #31504 since `from_breaks`/`from_arrays` is smart enough to infer `pd.NaT` as `interval[datetime64[ns]]`, so the failing example there will work.  This isn't perfect though because `interval[timedelta64[ns]]` will fail in a similar manner but I imagine that's a much rarer case.

Reply:
@jschendel it seems that `IntervalArray.from_breaks` converts NaT to NaN,

```python
In [12]: pd.arrays.IntervalArray.from_breaks([pd.NaT] * 4)
Out[12]:
<IntervalArray>
[nan, nan, nan]
Length: 3, closed: right, dtype: interval[datetime64[ns]]
```

Right now, I'd prefer to keep this fix targeted for fixing the regression, and follow up on the datetime / timedelta support.

Reply:
> it seems that IntervalArray.from_breaks converts NaT to NaN

All NA values have a string representation of `nan` in `IntervalArray`/`IntervalIndex` but behind the scenes things are stored correctly.  Notice that the dtype in your output is `interval[datetime64[ns]]`.  If you look at `left`/`right` they'd be `DatetimeIndex` with `NaT` values.  IIRC the logic here was that the non-NA elements of `interval[datetime64[ns]]` aren't actually times so having `NaT` didn't make sense.

> I'd prefer to keep this fix targeted for fixing the regression

Fair enough.

Reply:
Ah, good catch. If things are working correctly then I'll push a commit adding a test.